apps/financial-aid/web-veita/src/components/StateModal/StateModal.tsx

```diff
@@ -63,103 +68,125 @@ const StateModal = ({
-        >
-          <Box
-            paddingLeft={4}
-            paddingY={2}
-            background="blue400"
-            className={styles.modalHeadline}
-          >
-            <Text fontWeight="semiBold" color="white">
-              {headingText(inputType)}
-            </Text>
-          </Box>
-
-          <Box
-            display="flex"
-            className={cn({
-              [`${styles.container}`]: true,
-              [`${styles.showInput}`]: inputType.show,
-            })}
-          >
-            <OptionsModal
-              state={application.state}
-              onClick={(e, stateOption) => {
-                e.stopPropagation()
-                if (
-                  stateOption === ApplicationState.APPROVED ||
-                  stateOption === ApplicationState.REJECTED
-                ) {
-                  setInputType({
-                    show: !inputType.show,
-                    type: stateOption,
-                  })
-                } else {
-                  saveStateApplication(application, stateOption)
-                }
-              }}
-            />
-
-            <InputModal
-              onShowInputChange={(e) => {
-                e.stopPropagation()
-                setInputType({
-                  ...inputType,
-                  show: false,
-                })
-              }}
-              type={inputType.type}
-              onSaveState={(e, amount, comment) => {
-                e.stopPropagation()
-                if (inputType.type) {
-                  saveStateApplication(
-                    application,
-                    inputType.type,
-                    amount,
-                    comment,
-                  )
-                }
-              }}
-            />
-          </Box>
-        </Box>
+      <Box
+        paddingLeft={4}
+        paddingY={2}
+        background="blue400"
+        className={styles.modalHeadline}
+      >
+        <Text fontWeight="semiBold" color="white">
+          {headingText(selected)}
+        </Text>
+      </Box>
+
+      <Box display="block" width="full" padding={4}>
+        <AnimateSharedLayout type="crossfade">
+          <OptionsModal
+            isModalVisable={selected === undefined}
+            activeState={currentState}
+            onClick={(e, stateOption) => {
+              e.stopPropagation()
+
+              const goToNextWindow = stateNeedInput.find(
+                (item) => stateOption === item.state,
+              )
+              if (goToNextWindow) {
+                setSelected(stateOption)
+                return
+              }
+
+              saveStateApplication(applicationId, stateOption)
+            }}
+          />
+
+          <DataNeededModal
+            isModalVisable={selected === ApplicationState.DATANEEDED}
+            onCancel={onClickCancel}
+            onSaveApplication={(comment?: string) => {
+              if (!comment || !selected) {
+                setSelected(undefined)
```

Comment:
Should we add a error message and display it to the user instead of setting selected to undefined?

And should the modal take care of this logic?